File: 1038-binary-search-tree-to-greater-sum-tree/1038-binary-search-tree-to-greater-sum-tree.cpp
using namespace std;

int recursive(TreeNode* root,int& val) {
    if (root == nullptr) {
        return 0;
    }
    recursive(root->right,val);
    val += root->val;
    root->val = val;
    recursive(root->left,val);
    return 0;
}

class Solution {
public:
    TreeNode* bstToGst(TreeNode* root) {
        int val = 0;
        recursive(root,val);
        return root;
    }
};

File: 1038-binary-search-tree-to-greater-sum-tree/1038-binary-search-tree-to-greater-sum-tree.java
class Solution {
    public TreeNode bstToGst(TreeNode root) {
        dfs(root);
        return root;
        }

    public void dfs(TreeNode root){
        Stack<TreeNode> s = new Stack<>();
        s.push(root);
        HashMap<TreeNode,Boolean> visited = new HashMap<>();
        int sum = 0;
        while(!s.empty()){
            TreeNode curr = s.pop();
            if (!visited.containsKey(curr)){
                visited.put(curr,true);
                if (curr.left != null){
                    s.push(curr.left);
                }
                s.push(curr);
                if (curr.right != null){
                    s.push(curr.right);
                }
            }else{
                sum += curr.val;
                curr.val = sum ;
            }
        }
    }
}